src/Frontend/Core.hs

```diff
@@ -247,13 +251,15 @@ redirectFormHandler getPage processor = getH :<|> postH
         v <- getForm fa (processor1 page)
         renderer page v fa
 
-    postH env = do
+    postH formData = do
         page <- getPage
         let fa = absoluteUriPath $ formAction page
+            env = getFormDataEnv formData
         (v, mpayload) <- postForm fa (processor1 page) (\_ -> return $ return . runIdentity . env)
-        case mpayload of
+        (case mpayload of
             Just payload -> processor2 page payload >>= redirect
-            Nothing      -> renderer page v fa
+            Nothing      -> renderer page v fa)
+            `finally` cleanupTempCsvFiles formData
```

Comment:
Indeed, seems like a good place to clean up. But we need to be sure the CSV files are already fully read by this point (lazy IO). I can't deduce it from the type of the `processor` argument, so I'd add a contract to the haddock ("The @processor@ argument should read any opened files eagerly, because they are closed at the end of the call"). Fortunately, I can see the code of our processor for CSV files *is* strict enough, because it parses the whole file to decide whether to throw an error. We should be fine as long as we don't let users upload files for future use (e.g., if they upload an avatar, we should display it while still serving the same request or at least calculate its size, or something).